### PR TITLE
Enable SoundSourceFFmpeg as a fallback for broken FAAD v2.9.1

### PR DESCRIPTION
**Pull request by uklotzde**
SoundSourceM4A fails to decode any audio data after successfully opening the file. Mixxx loads .m4a files but then displays an empty waveform and the audio stream becomes empty after trying to decode it.

This PR adds a basic read test to verify that the audio stream is actually readable. Otherwise sound sources with a lower priority get the chance to open the file. This wasn't possible before.

Tested with Fedora 32. RPMFusion distributes the broken FAAD v2.9.1 library like Arch does.

A backport to 2.2.x would be nice, but requires too much work. If anyone feels responsible to do this feel free to take this commit as a starting point. I will not invest any time to do this, neither for the implementation **nor for a review**!

Links:
https://www.mixxx.org/forums/viewtopic.php?f=3&t=13102
https://www.mixxx.org/forums/viewtopic.php?f=3&t=13157

### CODE REVIEW

**Comment by daschuer** [comment]
Here is the upstream discussion for reference. 
https://github.com/knik0/faad2/commit/920ec985a74c6f88fe507181df07a0cd7e51d519
https://github.com/knik0/faad2/commit/a8dc3f8ce67f4069cfa4d5cf0fcc2c6e8ef2c2aa

**Comment by daschuer** [comment]
I am afraid the issue will spread, because Ubuntu Focal seems to be effected as well: 
https://packages.ubuntu.com/focal/libfaad2

**Comment by daschuer** [comment]
Can one confirm the issue on Ubuntu? 
Maybe the package maintainer can help.  
@fabiangreffrath Can you?

File: src/sources/audiosource.cpp
Comment (daschuer) [inline src/sources/audiosource.cpp:189]: "the" or "some"?

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp:502]: This warning is redundant, because verifyReadable() above warns as well.

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp:515]: This log message is redundant if we not return in the if condition below.

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp:521]: I think we can improve readability by using 
} else if (openResult == mixxx::SoundSource::OpenResult::Failed) {
...
} else {
    DEBUG_ASSERT(openResult == mixxx::SoundSource::OpenResult::Abort)
...

https://github.com/mixxxdj/mixxx/blob/9d86357e736e19a5a6ec6d3af5107d492b5e10c1/src/sources/audiosource.h#L176

Should mention the second path

**Review by daschuer — CHANGES_REQUESTED** [review]
This is a good step for reliability. 

Unfortunately it creates a bad situation for the current problem, because most likely faad and ffmpeg will have different offsets. So this will mess up the beat-grids after upgrading to an effected distro and will mess up the beat grids again after the issue has been fixed.

**Comment by daschuer** [comment]
@uklotzde can you verify if there is an offset issue? 

How does this relate to the deprecation of libmp4v2
https://bugs.launchpad.net/mixxx/+bug/1842208

File: src/sources/soundsourceproxy.cpp
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp:554]: Is this one necessary?
We have
* Opening File...
* Failed to open...
* Opening File...
* Failed to open...
* Unable to decode...

I would remove it.
Or rename it to "Give up"

A log improvement could be to count the attempts.

This helps to distinguish automatic attempts from user attempts.

**Review by daschuer — CHANGES_REQUESTED** [review]

**Comment by daschuer** [comment]
Do you have some insights about the topics I have raised above?

How many files are effected?
Is the an offset issue?

**Comment by uklotzde** [comment]
No, I haven't checked the offsets and I cannot make general statements. I'm using FFmpeg and have already upgraded to Fedora 32 where FAAD2 is broken. I need to build in a VM for verifying a few files. The offset may vary for different iTunes encoder versions and I am not able to test all possible variations.

The outcome of this investigation doesn't matter for this PR. The fallback solution doesn't only affects file formats. The decision about how to deal with the FAAD2 situation is a separate topic.

File: src/sources/soundsourceproxy.cpp
Comment (uklotzde) [reply src/sources/soundsourceproxy.cpp:545]: I will not rewrite this entire part. Log messages have been improved now. Finished here.

**Comment by daschuer** [comment]
> The decision about how to deal with the FAAD2 situation in particular is a separate topic.

Yes, under normal circumstance you are right. But we should be aware what it means for the user. 
I think I can perform the test here. Do you have broken files or are all of them broken?

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
This LGTM, but we should hold it of until we know what will happen on the users system.

**Comment by uklotzde** [comment]
I have tested with a single file from the iTunes Store. No visible or audible offsets when switching from FAAD2 to FFmpeg.

While testing and copying the Mixxx library from the VM and relinking the music folder to a different location I discovered a wrong debug assertion. Fixed.

**Comment by uklotzde** [comment]
Users are currently unable to play their .m4a files! I don't see any reason why we should hold this back??

**Comment by daschuer** [comment]
> No visible or audible offsets when switching from FAAD2 to FFmpeg.

Good to know. So we have nothing else to do here. Thank you for testing.

**Comment by daschuer** [comment]
Waiting for CI

**Comment by Holzhaus** [comment]
In case there are offsets, we could just switch to FFmpeg for M4A unconditionally. That would allow us to update all positions for M4A files in the database via a schema migration.

**Comment by Holzhaus** [comment]
> Do you have broken files or are all of them broken?

Converted a WAV file to M4A via FFmpeg:

    ffmpeg -i somefile.wav somefile.m4a

I can't play this file with FAAD.

**Comment by daschuer** [comment]
CI failure is a timout.

**Comment by daschuer** [comment]
LGTM

**Comment by fabiangreffrath** [comment]
Could you guys please check if this issue persists with FAAD2 2.9.2 which I have just released yesterday?

**Comment by uklotzde** [comment]
@fabiangreffrath I have installed the [RPMFusion package for f32](http://koji.rpmfusion.org/koji/buildinfo?buildID=15113) and still get known error "AAC decoding error: 21 Unexpected channel configuration change".

Decoding manually using faad works though!?

**Comment by fabiangreffrath** [comment]
> Decoding manually using faad works though!?

Is this an actual question?

**Comment by Piscium** [comment]
> Could you guys please check if this issue persists with FAAD2 2.9.2 which I have just released yesterday?

Issue still exists with FAAD2 2.9.2 and Mixxx 2.2.4 in Arch. I have thousands of files with this issue that cannot played in Mixxx but play just fine in Audacious, VLC and Quodlibet.

**Comment by fabiangreffrath** [comment]
But doesn't VLC also use faad2 internally? Though, they apply a couple of patches. Could one of them make the difference? 

https://github.com/videolan/vlc/tree/master/contrib/src/faad2

**Comment by Piscium** [comment]
> But doesn't VLC also use faad2 internally? Though, they apply a couple of patches. Could one of them make the difference?

I am playing a m4a with VLC as I write this. If I press Ctrl J to go to the Current Media Information window, Codec tab, it says:
Codec: MPEG AAC Audio (mpp4a)

I don't know if that is FAAD or some other codec.

**Comment by uklotzde** [comment]
VLC seems to be using a workaround for this bug: https://github.com/videolan/vlc/blob/b3fe59fad88112045e8c5ddda325fb3a06a2843e/modules/codec/faad.c#L354

@fabiangreffrath Is this a bug or intended behavior? It didn't occur with all versions up to 2.8.8.

**Comment by Piscium** [comment]
Btw, I did a quick test of the git Mixxx master using the Arch AUR git package and it works fine with respect to this issue.

**Comment by fabiangreffrath** [comment]
@uklotzde The code was changed to fix a vulnerability. I guess the library is more strict regarding standards compliance now, but this leads to compatibility issues. 

Please note that I am not the real maintainer for faad2. I am just the last person alive left with commit rights to the repository.

**Comment by fabiangreffrath** [comment]
Could someone please point me to a sample file that *should* play (and played before FAAD 2.8.8) but doesn't play now anymore?

**Comment by uklotzde** [comment]
Example for decoding error 21: https://raw.githubusercontent.com/robovm/apple-ios-samples/master/avTouch/sample2ch.m4a

**Comment by fabiangreffrath** [comment]
Thank you, but I cannot seem to reproduce the issue with this file and FAAD 2.9.2:
```
./frontend/faad.exe ../sample2ch.m4a -o out.wav
 *********** Ahead Software MPEG-4 AAC Decoder V2.9.2 ******************

 Build: Jun 25 2020
 Copyright 2002-2004: Ahead Software AG
 http://www.audiocoding.com
 bug tracking: https://sourceforge.net/p/faac/bugs/
 Floating point version

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License.

 **************************************************************************

**** MP4 header ****
Brand:                  mp42(version 1)
Compatible brands:      mp42mp41
*track media type: 'soun': OK
Modification Time:              Fri Feb 04 23:31:11 2011

Samplerate:             48000
Total samples:          971776
Total channels:         2
Bits per sample:        16
Buffer size:            6144
Max bitrate:            192000
Average bitrate:        192000
Samples per frame:      0
Frames:                 949
ASC size:               2
Duration:               20.2 sec
Data offset/size:       16f9/0
********************
parse error: atom 'udta' not found
can't read atom name @578fb
../sample2ch.m4a file info:

LC AAC  20.245 secs, 2 ch, 48000 Hz

  ---------------------
 | Config:  2 Ch       |
  ---------------------
 | Ch |    Position    |
  ---------------------
 | 00 | Left front     |
 | 01 | Right front    |
  ---------------------

Decoding ../sample2ch.m4a took:  0.08 sec. 259.56x real-time.

```

**Comment by Piscium** [comment]
I am on Linux so cannot easily run a Windows exe. This is the sort of file I had problems with, I have thousands of files like this. It is encrypted and  the password is
faadissue

After unencrypting add the extension m4a.

[sample.zip](https://github.com/mixxxdj/mixxx/files/4838690/sample.zip)

**Pull request merged**
